R/mrmash_wrapper.R

```diff
@@ -291,30 +273,66 @@ compute_w0 <- function(Bhat, ncomps) {
   return(w0)
 }
 
+
 ### Filter data-driven matrices
-filter_data_driven_mats <- function(Y, data_driven_mats) {
+filter_data_driven_mats <- function(Y, data_driven_mats, data_driven_mats_w = NULL, prior_weights_min = 1e-4) {
   conditions_to_keep <- colnames(Y)
-
   # Check if colnames of Y is a subset of column names of each element in data_driven_mats
-  for (mat_name in names(data_driven_mats)) {
-    mat <- data_driven_mats[[mat_name]]
+  data_driven_mats <- lapply(data_driven_mats, function(mat, to_keep) {
     missing_conditions <- setdiff(conditions_to_keep, colnames(mat))
-
     if (length(missing_conditions) > 0) {
-      stop(paste(
-        "Condition(s)", paste(missing_conditions, collapse = ", "),
-        "not found in matrix", mat_name
-      ))
+      stop(paste("Condition(s)", paste(missing_conditions, collapse = ", "), "not found in matrix", mat_name))
+    }
+    mat[to_keep, to_keep]
+  }, conditions_to_keep)
+  
+  for (mat_name in names(data_driven_mats)) {
+    # remove null component 
+    if (all(data_driven_mats[[mat_name]] == 0)) {
+      data_driven_mats[[mat_name]] <- NULL
+      if (!is.null(data_driven_mats_w)) data_driven_mats_w <- data_driven_mats_w[!names(data_driven_mats_w) %in% mat_name]
+      next
+    }
+    if(!is.null(data_driven_mats_w)){
+      if(data_driven_mats_w[mat_name] < prior_weights_min) {
+        data_driven_mats_w <- data_driven_mats_w[!names(data_driven_mats_w) %in% mat_name]
+        data_driven_mats[[mat_name]] <- NULL
+      }
+      next
     }
   }
+  message(paste(length(data_driven_mats), "components of data driven matrices remained after filtering. "))
+  return(list(U=data_driven_mats, w=data_driven_mats_w))
+}
 
-  data_driven_mats_filt <- lapply(data_driven_mats, function(x, to_keep) {
-    x[to_keep, to_keep]
-  }, conditions_to_keep)
 
-  return(data_driven_mats_filt)
+### Re-normalize mrmash weight w0 for total weight sum to 1 
+#' @param w0 is the weight of mr.mash prior matrices that was generated from mr.mash() function. 
+rescale_cov_w0 <- function(w0){
+    # remove null component 
+    w0 <- w0[names(w0) != "null"]
```

Comment:
Are you sure that here the null component will always have the name `null`? If not you need to use a more robust way.

Reply:
I think this `null` component is introduced at the step of `S0 <- expand_covs(S0_raw, prior_grid, zeromat = TRUE)`  and we did not provide users any options to change parameters in this step, the null component should be always named as `null`. Please confirm @fmorgante

Reply:
sounds good then!